### PR TITLE
Container issue node and description panel [ROAD-451]

### PR DESCRIPTION
**Pull request by ArtsiomCh**
image

Also refactoring of description panels (OSS, IaC) UI to extract some common parts.

### CODE REVIEW

**Comment by bastiandoetsch** [comment]
Please review the layout with @Andy Gongea - the Introduced through part looks unexpected to me.

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/LabelProvider.kt [hunk 1]
```diff
@@ -34,9 +34,9 @@ class LabelProvider {
         return createLinkLabel(URL("$cweBaseUrl/${cwe.removePrefix("CWE-")}.html"), cwe)
     }
 
-    fun getVulnerabilityLabel(id: String): JLabel {
-
-        return createLinkLabel(URL("$vulnerabilityBaseUrl/$id"), id)
+    fun getVulnerabilityLabel(id: String, idUrl: String? = null): JLabel {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/toolwindow/LabelProvider.kt:37]: Missing a unit test for this

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/LabelProvider.kt [hunk 2]
```diff
@@ -49,6 +49,8 @@ class LabelProvider {
 
     fun createLinkLabel(url: URL, text: String): LinkLabel<*> {
         val openLinkAction = OpenLinkAction(url)
-        return LinkLabel.create(text, openLinkAction)
+        return LinkLabel.create(text, openLinkAction).apply {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/toolwindow/LabelProvider.kt:52]: This is not correct - the toolTipText should be provided as a parameter, as the link could be something else  than a description.

File: src/main/kotlin/io/snyk/plugin/ui/UIUtils.kt [hunk 3]
```diff
@@ -164,16 +167,119 @@
     colSpan: Int = 1,
     anchor: Int = GridConstraints.ANCHOR_CENTER,
     fill: Int = GridConstraints.FILL_NONE,
-    hSizePolicy: Int = GridConstraints.SIZEPOLICY_FIXED,
-    vSizePolicy: Int = GridConstraints.SIZEPOLICY_FIXED,
+    HSizePolicy: Int = GridConstraints.SIZEPOLICY_FIXED,
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/UIUtils.kt:170]: in our own methods, I would like to see compliance with kotlin/java naming guidelines - any reason, you renamed it to have a Capitalized beginning?

File: src/main/kotlin/io/snyk/plugin/ui/UIUtils.kt [hunk 4]
```diff
@@ -164,16 +167,119 @@
...
+        row, column, rowSpan, colSpan, anchor, fill, HSizePolicy, VSizePolicy, minimumSize, preferredSize,
         maximumSize, indent, useParentLayout
     )
 }
+
+fun baseGridConstraintsAnchorWest(
+    row: Int,
+    column: Int = 0,
+    fill: Int = GridConstraints.FILL_NONE,
+    HSizePolicy: Int = GridConstraints.SIZEPOLICY_FIXED,
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/UIUtils.kt:188]: same

File: src/main/kotlin/snyk/container/ui/BaseImageRemediationDetailPanel.kt [hunk 5]
```diff
@@ -41,8 +43,7 @@
         this.add(
             Spacer(),
             baseGridConstraints(
-                9,
-                anchor = GridConstraints.ANCHOR_CENTER,
+                row = 9,
```
Comment (bastiandoetsch) [inline src/main/kotlin/snyk/container/ui/BaseImageRemediationDetailPanel.kt:46]: this adds so much to readability - love it.

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+        this.name = "ContainerIssueDetailPanel"
+    }
+
+    private fun titlePanel(): JPanel {
+        val titlePanel = JPanel()
+        titlePanel.layout = GridLayoutManager(2, 1, Insets(0, 0, 0, 0), -1, 5)
+
+        val titleLabel = JLabel().apply {
+            font = io.snyk.plugin.ui.getFont(Font.BOLD, 20, font)
```
Comment (bastiandoetsch) [inline src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt:63]: The fontsize 20 should be a constant :)

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        cwes = issue.identifiers?.cwe ?: emptyList(),
+        cves = issue.identifiers?.cve ?: emptyList(),
+        cvssScore = issue.cvssScore,
+        cvsSv3 = issue.CVSSv3,
+        id = issue.id
+    )
+
+    private fun mainPanel(): JPanel {
+        val panel = JPanel(
+            GridLayoutManager(11, 2, Insets(20, 0, 20, 0), 50, -1)
```
Comment (bastiandoetsch) [inline src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt:99]: As it's a new class, I'm gonna point it out... Inset numbers should be constants describing the spacing/padding that is covered by them (e.g. TOP_PADDING=20)

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+import javax.swing.JEditorPane
+import javax.swing.JLabel
+import javax.swing.JPanel
+
+class ContainerIssueDetailPanel(
+    private val issue: ContainerIssue
+) : JPanel() {
+
+    init {
+        this.layout = GridLayoutManager(10, 1, Insets(20, 10, 20, 20), -1, 10)
```
Comment (bastiandoetsch) [inline src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt:27]: the numbers should be constants describing what it does

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        id = issue.id
+    )
+
+    private fun mainPanel(): JPanel {
+        val panel = JPanel(
+            GridLayoutManager(11, 2, Insets(20, 0, 20, 0), 50, -1)
+        )
+
+        fun boldLabel(text: String) =
+            JLabel(text).apply { font = io.snyk.plugin.ui.getFont(Font.BOLD, -1, JLabel().font) }
```
Comment (bastiandoetsch) [inline src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt:103]: what does -1 do as a font size? Using a constant would make it easier to understand.

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        if (introducedThrough.isEmpty()) {
+            return null
+        }
+        val introducedThroughAsHtml: String = introducedThrough.joinToString(separator = ", ") { it }
+
+        return getReadOnlyClickableHtmlJEditorPane(introducedThroughAsHtml)
+    }
+
+    private fun overviewPanel(): JPanel {
+        val panel = JPanel(GridLayoutManager(2, 2, Insets(0, 5, 0, 0), -1, 0))
```
Comment (bastiandoetsch) [inline src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt:142]: see above

File: src/main/kotlin/snyk/container/ContainerIssues.kt [hunk 11]
```diff
@@ -33,6 +33,7 @@
     val severity: String,
     val identifiers: Identifiers? = null,
     val cvssScore: String? = null,
+    var CVSSv3: String? = null,
```
Comment (bastiandoetsch) [inline src/main/kotlin/snyk/container/ContainerIssues.kt:36]: old version was better, leading a field/attribute with a capitalized letter is always confusing to me

Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/ContainerIssues.kt:36]: and if it's about automatic conversion - let's use the @SerializedName annotation

**Comment by bastiandoetsch** [comment]
Nice start to getting the UI code at least a bit better!

**Review by bastiandoetsch — APPROVED** [review]
Regarding the improvement requests - I'll approve the PR now, so you're not blocked by me once you've worked on them.

File: src/main/kotlin/io/snyk/plugin/ui/UIUtils.kt (hunk 3 above)
Comment (ArtsiomCh) [reply src/main/kotlin/io/snyk/plugin/ui/UIUtils.kt:145]: to easily map (visually) it to the underlying method params. But ok, lets follow camelCase here.

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt (hunk 6 above)
Comment (ArtsiomCh) [reply src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt:63]: fixed during other refactoring

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt (hunk 7 above)
Comment (ArtsiomCh) [reply src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt]: let's leave it as is for now. then it consistent across all Description panels. 
that definitely need refactoring later. But for now we gonna copy-past it anyway. Also Idea helps :) with hints enabled:
image

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt (hunk 8 above)
Comment (ArtsiomCh) [reply src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt:27]: see reply above

File: src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt (hunk 9 above)
Comment (ArtsiomCh) [reply src/main/kotlin/snyk/container/ui/ContainerIssueDetailPanel.kt]: fixed during other refactoring

**Comment by ArtsiomCh** [comment]
Updated OSS panel by Andy's comments:
image

**Comment by bastiandoetsch** [comment]
LGTM

**Pull request merged**